### PR TITLE
feat(schema-registry): add methods for exporting and importing all data from DB to json

### PR DESCRIPTION
**Pull request by frondeus**
TODO:
* [ ] Documentation
* [ ] Manual tests

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,6 +1,4 @@
 registry_db/
 storage_db/
 target/
-.vscode/
```
Comment (USSAMOH2) [inline .gitignore:4]: Why would we want these things committed?

File: schema-registry/src/error.rs [hunk 2]
```diff
@@ -54,8 +54,13 @@ impl From<jsonschema::CompilationError> for RegistryError {
 pub enum MalformedError {
     #[error("Schema in database was malformed with id {0}")]
     MalformedSchema(Uuid),
+
```
Comment (USSAMOH2) [inline schema-registry/src/error.rs:57]: What's with the extra space? I personally think we don't need it, but if you would prefer to have it, we should be consistent across all of our error enums.

File: schema-registry/src/rpc.rs [hunk 3]
```diff
@@ -59,6 +60,16 @@ impl SchemaRegistryImpl {
             .unwrap_or_else(abort_on_poison)
             .replicate_message(event)
     }
+
+    pub fn export_all(&self) -> anyhow::Result<DbExport> {
+        let result = self.db.export_all()?;
+        Ok(result)
```
Comment (USSAMOH2) [inline schema-registry/src/rpc.rs:66]:
```suggestion
        self.db.export_all().context("Failed to export the entire database")
```

File: schema-registry/src/rpc.rs [hunk 4]
```diff
@@ -59,6 +60,16 @@ impl SchemaRegistryImpl {
...
+    }
+
+    pub fn import_all(&self, imported: DbExport) -> anyhow::Result<()> {
+        self.db.import_all(imported)?;
+        Ok(())
```
Comment (USSAMOH2) [inline schema-registry/src/rpc.rs:71]:
```suggestion
        self.db.import_all(imported).context("failed to import database")
```

File: schema-registry/src/types/storage/edges.rs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    fn from_properties(properties: EdgeProperties) -> Option<Self> {
+        let schema_id = properties.edge.key.outbound_id;
+        let view_id = properties.edge.key.inbound_id;
+        Some(Self { schema_id, view_id })
+    }
+
+    fn db_type() -> Type {
+        SCHEMA_VIEW_EDGE_TYPE.clone()
```
Comment (USSAMOH2) [inline schema-registry/src/types/storage/edges.rs:36]: Do you need the `lazy_static!` definition if you only really need them here?

File: schema-registry/src/types/storage/edges.rs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct SchemaDefinition {
+    pub schema_id: Uuid,
+    pub def_id: Uuid,
```
Comment (USSAMOH2) [inline schema-registry/src/types/storage/edges.rs:43]: I know some people prefer brevity, but I would say that the elision here from `definition` to `def` makes the code less readable. How strongly do you prefer the shorter version?

File: schema-registry/src/types.rs [hunk 7]
```diff
@@ -79,3 +80,35 @@ impl VersionedUuid {
...
+#[derive(Default, Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct DbExport {
+    pub schemas: HashMap<Uuid, vertices::Schema>,
+    pub definitions: HashMap<Uuid, vertices::Definition>,
+    pub views: HashMap<Uuid, vertices::View>,
+    pub def_edges: Vec<edges::SchemaDefinition>,
+    pub view_edges: Vec<edges::SchemaView>,
+}
+
+fn get_vertex_property_or<T: DeserializeOwned>(
```
Comment (USSAMOH2) [inline schema-registry/src/types.rs:94]:
```suggestion
fn extract_vertex_property<T: DeserializeOwned>(
```
To imply that we are removing the desired property.

File: schema-registry/src/types.rs [hunk 8]
```diff
@@ -79,3 +80,35 @@ impl VersionedUuid {
...
+    name: &'static str,
+) -> Option<T> {
+    properties
+        .props
+        .drain_filter(|prop| prop.name == name)
+        .next()
+        .and_then(|prop| serde_json::from_value(prop.value).ok())
+}
+
+fn get_edge_property_or<T: DeserializeOwned>(
```
Comment (USSAMOH2) [inline schema-registry/src/types.rs:105]:
```suggestion
fn extract_edge_property<T: DeserializeOwned>(
```

**Review by USSAMOH2 — CHANGES_REQUESTED** [review]

File: schema-registry/src/types/storage/edges.rs [hunk 9]
```diff
@@ -0,0 +1,72 @@
+use super::*;
```
Comment (luke-biel) [inline schema-registry/src/types/storage/edges.rs:1]: Not a fan

File: .gitignore (hunk 1 above)
Comment (frondeus) [reply .gitignore:4]: We don't!
But as @luke-biel pointed, its committer responsibility to have these folders in his own personal global `.gitignore`.

File: schema-registry/src/error.rs (hunk 2 above)
Comment (frondeus) [reply schema-registry/src/error.rs:57]: Well for me its easier to read

File: schema-registry/src/types/storage/edges.rs (hunk 5 above)
Comment (frondeus) [reply schema-registry/src/types/storage/edges.rs:36]: As far as I understand `Type::new()` has some validation which we unwrap in lazy_static.
I could use constructor here and forget about it but then we would validate on every callsite.
I we don't care about it then yeah, we can remove lazy_static.

Comment (luke-biel) [reply schema-registry/src/types/storage/edges.rs:43]: We can just create Type with unchecked, we are guaranteed these are correct.

**Comment by luke-biel** [comment]
@frondeus docs aren't as important as other parts of this commit, so you can add them in separate PR.

File: schema-registry/src/types/storage/edges.rs (hunk 5 above)
Comment (frondeus) [reply schema-registry/src/types/storage/edges.rs:43]: But then we `.clone()` - there is allocation anyway right?

**Comment by frondeus** [comment]
> @frondeus docs aren't as important as other parts of this commit, so you can add them in separate PR.

I know, that's why i focus on other parts than docs.

File: schema-registry/src/types/storage/edges.rs (hunk 5 above)
Comment (USSAMOH2) [reply schema-registry/src/types/storage/edges.rs:43]: I'm fine with leaving it as is, because the `unchecked` alternative uses unsafe.

**Review by USSAMOH2 — APPROVED** [review]

**Pull request merged**